gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc

```diff
@@ -0,0 +1,245 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-toplevel-name-resolver-2.0.h"
+#include "rust-ast-full.h"
+#include "rust-hir-map.h"
+
+namespace Rust {
+namespace Resolver2_0 {
+
+TopLevel::TopLevel (Resolver &resolver) : DefaultResolver (resolver) {}
+
+template <typename T>
+void
+TopLevel::insert_or_error_out (const Identifier &identifier, const T &node,
+			       Namespace ns)
+{
+  auto loc = node.get_locus ();
+  auto node_id = node.get_node_id ();
+
+  // keep track of each node's location to provide useful errors
+  node_locations.emplace (node_id, loc);
+
+  auto result = resolver.insert (identifier, node_id, ns);
+
+  if (!result)
+    {
+      rich_location rich_loc (line_table, loc);
+      rich_loc.add_range (node_locations[result.error ().existing]);
+
+      rust_error_at (rich_loc, ErrorCode ("E0428"),
+		     "%qs defined multiple times",
+		     identifier.as_string ().c_str ());
+    }
+}
+
+void
+TopLevel::go (AST::Crate &crate)
+{
+  for (auto &item : crate.items)
+    item->accept_vis (*this);
+}
+
+void
+TopLevel::visit (AST::Module &module)
+{
+  // FIXME: Do we need to insert the module in the type namespace?
+
+  auto sub_visitor = [this, &module] () {
+    for (auto &item : module.get_items ())
+      item->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Module, module.get_node_id (), sub_visitor,
+		   module.get_name ());
+}
+
+static bool
+is_macro_export (AST::MacroRulesDefinition &def)
+{
+  for (const auto &attr : def.get_outer_attrs ())
+    if (attr.get_path ().as_string () == "macro_export")
+      return true;
+
+  return false;
+}
+
+void
+TopLevel::visit (AST::MacroRulesDefinition &macro)
+{
+  // we do not insert macros in the current rib as that needs to be done in the
+  // textual scope of the Early pass. we only insert them in the root of the
```

Comment:
`macro` definitions (which libcore uses for some built in macros) are inserted in module scopes,

Reply:
grrrrr you're right. I hate macro name resolution rules lol. thank you! I'll take care of this